src/App.tsx

```diff
@@ -36,35 +50,63 @@ function App(props: Props) {
       state.signer &&
       state.address &&
       roundId &&
-      (previousAddress === null || previousSigner === null)
+      (!previousAddress || !previousSigner)
     ) {
       const address = state.address;
       const message = `UMA Protocol one time key for round: ${roundId}`;
+      const hexMessage = web3.utils.toHex(message);
       const keyExists = signingKeys[address];
-      if (!keyExists || keyExists.roundMessage !== message) {
-        state.signer
-          .signMessage(message)
-          .then((msg) => {
-            const key = {} as {
-              publicKey: string;
-              privateKey: string;
-              roundMessage: string;
-            };
-
-            const privateKey = derivePrivateKey(msg);
-            const publicKey = recoverPublicKey(privateKey);
-            key.privateKey = privateKey;
-            key.publicKey = publicKey;
-            key.roundMessage = message;
-
-            setSigningKeys((prevKeys) => {
-              return { ...prevKeys, [address]: key };
+      let keysInStorage =
+        localStorage.getItem(SIGNING_KEYS_STORAGE_KEY) || ({} as StorageKeys);
+
+      let keyExistsInStorage = false;
+      if (typeof keysInStorage === "string") {
+        keysInStorage = JSON.parse(keysInStorage) as StorageKeys;
+
+        keyExistsInStorage = has(keysInStorage, `${hexMessage}.${address}`);
+      }
+
+      if (keyExistsInStorage) {
+        const loggedInKey = keysInStorage[hexMessage][address];
+        setSigningKeys((prevKeys) => {
+          return { ...prevKeys, [address]: loggedInKey };
+        });
+      } else {
+        if (!keyExists || keyExists.roundMessage !== message) {
```

Comment:
Similar to the question above, since we're using the message to look the value up in the mapping -- why do we need to store it inside the data blob and then re-check it here?

Reply:
I believe this check exists solely if the user is on the app transitioning between rounds.

It's sort of an edge case to indicate they need to re-sign.